src/Control/Effect/Resumable.hs

```diff
@@ -0,0 +1,115 @@
+--
+--   Note that since we can’t tell whether the type indices are equal, let alone what 'Ord' instance to use for them, the comparator passed to 'liftCompare' always returns 'EQ'. Thus, 'SomeError' is best used with type-indexed GADTs for the error type.
+--
+--   prop> (SomeError (Identity a) `compare` SomeError (Identity b)) == EQ
+--   prop> (SomeError (Const a) `compare` SomeError (Const b)) == (a `compare` b)
+instance Ord1 err => Ord (SomeError err) where
+  SomeError exc1 `compare` SomeError exc2 = liftCompare (const (const EQ)) exc1 exc2
+
+-- | Showing for 'SomeError' is determined by a 'Show1' instance for the error type.
+--
+--   Note that since we can’t tell what 'Show' instance to use for the type index, the functions passed to 'liftShowsPrec' always return the empty 'ShowS'. Thus, 'SomeError' is best used with type-indexed GADTs for the error type.
+--
+--   prop> show (SomeError (Identity a)) == "SomeError (Identity )"
+--   prop> show (SomeError (Const a)) == ("SomeError (Const " ++ showsPrec 11 a ")")
+instance (Show1 err) => Show (SomeError err) where
+  showsPrec d (SomeError err) = showsUnaryWith (liftShowsPrec (const (const id)) (const id)) "SomeError" d err
+
+
+-- | Run a 'Resumable' effect, returning uncaught errors in 'Left' and successful computations’ values in 'Right'.
+--
+--   prop> run (runResumable (pure a)) == Right @(SomeError Identity) @Int a
+runResumable :: Effectful sig m => Eff (ResumableH err m) a -> m (Either (SomeError err) a)
+runResumable = runResumableH . interpret
+
+newtype ResumableH err m a = ResumableH { runResumableH :: m (Either (SomeError err) a) }
+
+instance Effectful sig m => Carrier (Resumable err :+: sig) (ResumableH err m) where
+  gen a = ResumableH (gen (Right a))
+  alg = algE \/ (ResumableH . alg . handle (Right ()) (either (gen . Left) runResumableH))
+    where algE (Resumable err _) = ResumableH (gen (Left (SomeError err)))
+
+
+-- | Run a 'Resumable' effect, resuming uncaught errors with a given handler.
+--
+--   Note that this may be less efficient than defining a specialized carrier type and instance specifying the handler’s behaviour directly. Performance-critical code may wish to do that to maximize the opportunities for fusion and inlining.
+--
+--   >>> data Err a where Err :: Int -> Err Int
+--
+--   prop> run (runResumableWith (\ (Err b) -> pure (1 + b)) (pure a)) == a
+--   prop> run (runResumableWith (\ (Err b) -> pure (1 + b)) (throwResumable (Err a))) == 1 + a
+runResumableWith :: (Carrier sig m, Monad m)
+                 => (forall x . err x -> m x)
+                 -> Eff (ResumableWithH err m) a
+                 -> m a
+runResumableWith with = runResumableWithH with . interpret
+
+newtype ResumableWithH err m a = ResumableWithH ((forall x . err x -> m x) -> m a)
+
+runResumableWithH :: (forall x . err x -> m x) -> ResumableWithH err m a -> m a
+runResumableWithH f (ResumableWithH m) = m f
```

Comment:
The `ResumableWithH` carrier takes a rank-_n_ higher-order function, and `flip` doesn’t play nicely with rank-_n_ constructors. Therefore, instead of using a record selector and flipping it, we just define the accessor as a function with the arguments in the order we want.